{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FFM\n",
    "FFM用のDatasetを作成し、kerasで実行する"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-01-12 06:10:03,607 utils 366 [INFO]    [logger_func] start \n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import datetime\n",
    "import glob\n",
    "import gc\n",
    "import os\n",
    "\n",
    "#========================================================================\n",
    "# Args\n",
    "#========================================================================\n",
    "key = 'card_id'\n",
    "target = 'target'\n",
    "ignore_list = [key, target, 'merchant_id', 'first_active_month']\n",
    "\n",
    "win_path = f'../features/4_winner/*.gz'\n",
    "fname=''\n",
    "# submit = pd.read_csv('../input/sample_submission.csv')\n",
    "submit = []\n",
    "\n",
    "HOME = os.path.expanduser('~')\n",
    "\n",
    "sys.path.append(f\"{HOME}/kaggle/data_analysis/library/\")\n",
    "import utils\n",
    "from preprocessing import get_ordinal_mapping, get_dummies\n",
    "from utils import logger_func\n",
    "try:\n",
    "    if not logger:\n",
    "        logger=logger_func()\n",
    "except NameError:\n",
    "    logger=logger_func()\n",
    "\n",
    "start_time = \"{0:%Y%m%d_%H%M%S}\".format(datetime.datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 3/3 [00:00<00:00, 89.19it/s]\n"
     ]
    }
   ],
   "source": [
    "#========================================================================\n",
    "# Data Load\n",
    "base = utils.read_df_pkl('../input/base*')\n",
    "win_path_list = glob.glob(win_path)\n",
    "train_path_list = []\n",
    "test_path_list = []\n",
    "for path in win_path_list:\n",
    "    if path.count('train'):\n",
    "        train_path_list.append(path)\n",
    "    elif path.count('test'):\n",
    "        test_path_list.append(path)\n",
    "\n",
    "base_train = base[~base[target].isnull()].reset_index(drop=True)\n",
    "base_test = base[base[target].isnull()].reset_index(drop=True)\n",
    "train_feature_list = utils.parallel_load_data(path_list=train_path_list)\n",
    "test_feature_list = utils.parallel_load_data(path_list=test_path_list)\n",
    "train = pd.concat(train_feature_list, axis=1)\n",
    "train = pd.concat([base_train, train], axis=1)\n",
    "test = pd.concat(test_feature_list, axis=1)\n",
    "test = pd.concat([base_test, test], axis=1)\n",
    "\n",
    "y = train[[key, target]]\n",
    "train.drop(target, axis=1, inplace=True)\n",
    "test.drop(target, axis=1, inplace=True)\n",
    "train.fillna(train.median(), inplace=True)\n",
    "test.fillna(test.median(), inplace=True)\n",
    "\n",
    "# FFMは最後の列がラベルになる\n",
    "train.sort_index(axis=1, inplace=True)\n",
    "test.sort_index(axis=1, inplace=True)\n",
    "\n",
    "train = train.merge(y, how='inner', on=key)\n",
    "train.head()\n",
    "#========================================================================\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "from sklearn.base import BaseEstimator\n",
    "from keras.layers import Input, Embedding, Dense,Flatten, Activation, dot, add\n",
    "from keras.models import Model\n",
    "from keras.regularizers import l2 as l2_reg\n",
    "from keras import initializers\n",
    "import itertools\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "\n",
    "def make_batches(size, batch_size):\n",
    "    nb_batch = int(np.ceil(size/float(batch_size)))\n",
    "    return [(i*batch_size, min(size, (i+1)*batch_size)) for i in range(0, nb_batch)]\n",
    "\n",
    "\n",
    "def batch_generator(X,y,batch_size=128,shuffle=True):\n",
    "    sample_size = X[0].shape[0]\n",
    "    index_array = np.arange(sample_size)\n",
    "    while 1:\n",
    "        if shuffle:\n",
    "            np.random.shuffle(index_array)\n",
    "        batches = make_batches(sample_size, batch_size)\n",
    "        for batch_index, (batch_start, batch_end) in enumerate(batches):\n",
    "            batch_ids = index_array[batch_start:batch_end]\n",
    "            X_batch = [X[i][batch_ids] for i in range(len(X))]\n",
    "            y_batch = y[batch_ids]\n",
    "            yield X_batch,y_batch\n",
    "\n",
    "\n",
    "def test_batch_generator(X,y,batch_size=128):\n",
    "    sample_size = X[0].shape[0]\n",
    "    index_array = np.arange(sample_size)\n",
    "    batches = make_batches(sample_size, batch_size)\n",
    "    for batch_index, (batch_start, batch_end) in enumerate(batches):\n",
    "        batch_ids = index_array[batch_start:batch_end]\n",
    "        X_batch = [X[i][batch_ids] for i in range(len(X))]\n",
    "        y_batch = y[batch_ids]\n",
    "        yield X_batch,y_batch\n",
    "\n",
    "\n",
    "def predict_batch(model,X_t,batch_size=128):\n",
    "    outcome = []\n",
    "    for X_batch,y_batch in test_batch_generator(X_t,np.zeros(X_t[0].shape[0]),batch_size=batch_size):\n",
    "        outcome.append(model.predict(X_batch,batch_size=batch_size))\n",
    "    outcome = np.concatenate(outcome).ravel()\n",
    "    return outcome\n",
    "\n",
    "\n",
    "\n",
    "def build_model(input_len, max_features,K=8,solver='adam',l2=0.0,l2_fm = 0.0):\n",
    "\n",
    "    inputs = []\n",
    "    flatten_layers=[]\n",
    "    columns = range(len(max_features))\n",
    "    for c in columns:\n",
    "        inputs_c = Input(shape=(1,), dtype='int32',name = 'input_%s'%c)\n",
    "        num_c = max_features[c]\n",
    "\n",
    "        embed_c = Embedding(\n",
    "                        input_dim=num_c, # 埋め込む特徴の次元\n",
    "                        output_dim=K, # 何次元に埋め込むか\n",
    "                        input_length=1,\n",
    "#                         input_length=1,\n",
    "                        name = 'embed_%s'%c,\n",
    "                        W_regularizer=l2_reg(l2_fm)\n",
    "                        )(inputs_c)\n",
    "\n",
    "              \n",
    "        flatten_c = Flatten()(embed_c)\n",
    "\n",
    "        inputs.append(inputs_c)\n",
    "        flatten_layers.append(flatten_c)\n",
    "\n",
    "    fm_layers = []\n",
    "\n",
    "    for emb1,emb2 in itertools.combinations(flatten_layers, 2):\n",
    "        \n",
    "#         dot_layer = merge([emb1,emb2], mode='dot', dot_axes=1)\n",
    "        dot_layer = dot(inputs=[emb1, emb2], axes=1)\n",
    "        \n",
    "        fm_layers.append(dot_layer)\n",
    "\n",
    "        \n",
    "    for c in columns:\n",
    "        num_c = max_features[c]\n",
    "        \n",
    "        embed_c = Embedding(\n",
    "                        num_c,\n",
    "                        1,\n",
    "                        input_length=1,\n",
    "#                         input_length=input_len,\n",
    "                        name = 'linear_%s'%c,\n",
    "                        W_regularizer=l2_reg(l2)\n",
    "                        )(inputs[c])\n",
    "\n",
    "        flatten_c = Flatten()(embed_c)\n",
    "\n",
    "        fm_layers.append(flatten_c)\n",
    "        \n",
    "#     flatten = merge(fm_layers, mode='sum')\n",
    "    flatten = add(fm_layers) \n",
    "    outputs = Activation('sigmoid',name='outputs')(flatten)\n",
    "    \n",
    "    model = Model(input=inputs, output=outputs)\n",
    "\n",
    "    model.compile(\n",
    "                optimizer=solver,\n",
    "                loss= 'binary_crossentropy'\n",
    "              )\n",
    "\n",
    "    return model\n",
    "\n",
    "\n",
    "class KerasFM(BaseEstimator):\n",
    "    def __init__(self, input_len, max_features=[], K=8, solver='adam', l2=0.0, l2_fm=0.0):\n",
    "        self.model = build_model(input_len, max_features,K,solver,l2=l2,l2_fm = l2_fm)\n",
    "\n",
    "    def fit(self, X, y, batch_size=128, nb_epoch=10, shuffle=True, verbose=1, validation_data=None):\n",
    "        self.model.fit(X,y,batch_size=batch_size,nb_epoch=nb_epoch,shuffle=shuffle,verbose=verbose,validation_data=None)\n",
    "\n",
    "    def fit_generator(self,X,y,batch_size=128,nb_epoch=10,shuffle=True,verbose=1,validation_data=None,callbacks=None):\n",
    "        tr_gen = batch_generator(X,y,batch_size=batch_size,shuffle=shuffle)\n",
    "        if validation_data:\n",
    "            X_test,y_test = validation_data\n",
    "            te_gen = batch_generator(X_test,y_test,batch_size=batch_size,shuffle=False)\n",
    "            nb_val_samples = X_test[-1].shape[0]\n",
    "        else:\n",
    "            te_gen = None\n",
    "            nb_val_samples = None\n",
    "\n",
    "        self.model.fit_generator(\n",
    "                tr_gen, \n",
    "                samples_per_epoch=X[-1].shape[0], \n",
    "                nb_epoch=nb_epoch, \n",
    "                verbose=verbose, \n",
    "                callbacks=callbacks, \n",
    "                validation_data=te_gen, \n",
    "                nb_val_samples=nb_val_samples, \n",
    "                max_q_size=10\n",
    "                )\n",
    "\n",
    "    def predict(self,X,batch_size=128):\n",
    "        y_preds = predict_batch(self.model,X,batch_size=batch_size)\n",
    "        return y_preds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Load & FFM Build"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1168, 36)\n",
      "(292, 36)\n",
      "(1459, 36)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_0\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_1\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_2\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_3\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_4\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_5\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_6\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_7\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_8\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_9\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_10\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_11\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_12\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_13\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_14\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_15\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_16\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_17\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_18\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_19\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_20\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_21\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_22\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_23\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_24\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_25\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_26\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_27\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_28\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_29\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_30\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_31\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_32\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_33\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_34\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:64: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=1168, output_dim=8, input_length=1, name=\"embed_35\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_0\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_1\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_2\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_3\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_4\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_5\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_6\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_7\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_8\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_9\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_10\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_11\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_12\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_13\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_14\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_15\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_16\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_17\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_18\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_19\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_20\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_21\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_22\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_23\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_24\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_25\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_26\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_27\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_28\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_29\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_30\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_31\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_32\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_33\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_34\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:92: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(1168, 1, input_length=1, name=\"linear_35\", embeddings_regularizer=<keras.reg...)`\n",
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:103: UserWarning: Update your `Model` call to the Keras 2 API: `Model(inputs=[<tf.Tenso..., outputs=Tensor(\"ou...)`\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "train = pd.read_csv('../../house_prise/input/train.csv')\n",
    "test = pd.read_csv('../../house_prise/input/test.csv')\n",
    "\n",
    "num_list = [col for col in train.columns if str(train[col].dtype).count('int') or str(train[col].dtype).count('float') ]\n",
    "\n",
    "train = train[num_list]\n",
    "num_list.remove('SalePrice')\n",
    "test = test[num_list]\n",
    "train.drop('Id', axis=1, inplace=True)\n",
    "test.drop('Id', axis=1, inplace=True)\n",
    "\n",
    "train, valid = train_test_split(train, test_size=0.2)\n",
    "tmp_y_train = train['SalePrice'].map(lambda x: 1 if x<120000 else 0)\n",
    "tmp_y_valid = valid['SalePrice'].map(lambda x: 1 if x<120000 else 0)\n",
    "train.drop('SalePrice', axis=1, inplace=True)\n",
    "valid.drop('SalePrice', axis=1, inplace=True)\n",
    "print(train.shape)\n",
    "print(valid.shape)\n",
    "print(test.shape)\n",
    "# sys.exit()\n",
    "\n",
    "max_features = [len(train[col]) for col in train.columns]\n",
    "model = KerasFM(input_len=len(train), max_features=max_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1168, 36)\n",
      "(1168, 1)\n",
      "(292, 36)\n",
      "(292, 1)\n",
      "36\n",
      "36\n",
      "36\n",
      "Epoch 1/10\n",
      " 128/1168 [==>...........................] - ETA: 0s - loss: 2.3788"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/anaconda3/envs/nn/lib/python3.6/site-packages/ipykernel_launcher.py:118: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1168/1168 [==============================] - 1s 989us/step - loss: 2.2970\n",
      "Epoch 2/10\n",
      "1168/1168 [==============================] - 1s 997us/step - loss: 2.2943\n",
      "Epoch 3/10\n",
      "1168/1168 [==============================] - 1s 994us/step - loss: 2.2942\n",
      "Epoch 4/10\n",
      "1168/1168 [==============================] - 1s 1ms/step - loss: 2.2943\n",
      "Epoch 5/10\n",
      "1168/1168 [==============================] - 1s 1000us/step - loss: 2.2941\n",
      "Epoch 6/10\n",
      "1168/1168 [==============================] - 1s 1ms/step - loss: 2.2941\n",
      "Epoch 7/10\n",
      "1168/1168 [==============================] - 1s 1000us/step - loss: 2.2940\n",
      "Epoch 8/10\n",
      "1168/1168 [==============================] - 1s 978us/step - loss: 2.2940\n",
      "Epoch 9/10\n",
      "1168/1168 [==============================] - 1s 992us/step - loss: 2.2939\n",
      "Epoch 10/10\n",
      "1168/1168 [==============================] - 1s 989us/step - loss: 2.2939\n"
     ]
    }
   ],
   "source": [
    "len_train = len(train)\n",
    "len_valid = len(valid)\n",
    "len_test = len(test)\n",
    "\n",
    "train_val = train.values\n",
    "y_train_val = tmp_y_train.values\n",
    "valid_val = valid.values\n",
    "y_valid_val = tmp_y_valid.values\n",
    "test_val = test.values\n",
    "\n",
    "x_train = train_val.reshape(len_train, 36)\n",
    "y_train = y_train_val.reshape(len_train, 1)\n",
    "x_valid = valid_val.reshape(len_valid, 36)\n",
    "y_valid = y_valid_val.reshape(len_valid, 1)\n",
    "x_test = test_val.reshape(len_test, 36)\n",
    "\n",
    "print(x_train.shape)\n",
    "print(y_train.shape)\n",
    "print(x_valid.shape)\n",
    "print(y_valid.shape)\n",
    "# (1168, 36)\n",
    "# (1168, 1)\n",
    "# (292, 36)\n",
    "# (292, 1)\n",
    "\n",
    "x_train = [i for i in x_train.T]\n",
    "x_valid = [i for i in x_valid.T]\n",
    "x_test = [i for i in x_test.T]\n",
    "\n",
    "print(len(x_train))\n",
    "print(len(x_valid))\n",
    "print(len(x_test))\n",
    "# 36\n",
    "# 36\n",
    "# 36\n",
    "\n",
    "model.fit(X=x_train, y=y_train, validation_data=(x_valid, y_valid))\n",
    "pred = model.predict(X=x_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "embedding_2 (Embedding)      (None, 36, 8)             8         \n",
      "_________________________________________________________________\n",
      "flatten_74 (Flatten)         (None, 288)               0         \n",
      "_________________________________________________________________\n",
      "dense_2 (Dense)              (None, 1)                 289       \n",
      "=================================================================\n",
      "Total params: 297\n",
      "Trainable params: 297\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Error when checking model input: the list of Numpy arrays that you are passing to your model is not the size the model expected. Expected to see 1 array(s), but instead got the following list of 36 arrays: [array([[50.],\n       [20.],\n       [20.],\n       ...,\n       [20.],\n       [50.],\n       [20.]]), array([[51.],\n       [80.],\n       [75.],\n       ...,\n       [40.],\n       [60.],\n       [98.]]), arr...",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-f14432780c80>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msummary\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m \u001b[0mhistotry\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepochs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbatch_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m32\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalidation_split\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/envs/nn/lib/python3.6/site-packages/keras/engine/training.py\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, **kwargs)\u001b[0m\n\u001b[1;32m    950\u001b[0m             \u001b[0msample_weight\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msample_weight\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    951\u001b[0m             \u001b[0mclass_weight\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mclass_weight\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 952\u001b[0;31m             batch_size=batch_size)\n\u001b[0m\u001b[1;32m    953\u001b[0m         \u001b[0;31m# Prepare validation data.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    954\u001b[0m         \u001b[0mdo_validation\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/nn/lib/python3.6/site-packages/keras/engine/training.py\u001b[0m in \u001b[0;36m_standardize_user_data\u001b[0;34m(self, x, y, sample_weight, class_weight, check_array_lengths, batch_size)\u001b[0m\n\u001b[1;32m    749\u001b[0m             \u001b[0mfeed_input_shapes\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    750\u001b[0m             \u001b[0mcheck_batch_axis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m  \u001b[0;31m# Don't enforce the batch size.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 751\u001b[0;31m             exception_prefix='input')\n\u001b[0m\u001b[1;32m    752\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    753\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0my\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/nn/lib/python3.6/site-packages/keras/engine/training_utils.py\u001b[0m in \u001b[0;36mstandardize_input_data\u001b[0;34m(data, names, shapes, check_batch_axis, exception_prefix)\u001b[0m\n\u001b[1;32m    100\u001b[0m                 \u001b[0;34m'Expected to see '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnames\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m' array(s), '\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    101\u001b[0m                 \u001b[0;34m'but instead got the following list of '\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 102\u001b[0;31m                 str(len(data)) + ' arrays: ' + str(data)[:200] + '...')\n\u001b[0m\u001b[1;32m    103\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnames\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    104\u001b[0m             raise ValueError(\n",
      "\u001b[0;31mValueError\u001b[0m: Error when checking model input: the list of Numpy arrays that you are passing to your model is not the size the model expected. Expected to see 1 array(s), but instead got the following list of 36 arrays: [array([[50.],\n       [20.],\n       [20.],\n       ...,\n       [20.],\n       [50.],\n       [20.]]), array([[51.],\n       [80.],\n       [75.],\n       ...,\n       [40.],\n       [60.],\n       [98.]]), arr..."
     ]
    }
   ],
   "source": [
    "from keras.models import Sequential\n",
    "from keras.layers import Flatten, Dense, Embedding\n",
    "\n",
    "model = Sequential()\n",
    "\n",
    "model.add(Embedding(1, 8, input_length=36))\n",
    "model.add(Flatten())\n",
    "model.add(Dense(1, activation='sigmoid'))\n",
    "model.compile(optimizer=\"rmsprop\", loss='binary_crossentropy', metrics=['acc'])\n",
    "model.summary()\n",
    "\n",
    "histotry = model.fit(x_train, y_train, epochs=10, batch_size=32, validation_split=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0],\n",
       "       [0],\n",
       "       [0],\n",
       "       ...,\n",
       "       [0],\n",
       "       [0],\n",
       "       [0]])"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_train"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nn",
   "language": "python",
   "name": "nn"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
